### PR TITLE
Optimize

### PR DESCRIPTION
**Pull request by TheMarex**
This needs a cleanup since there is also benchmarking code in there.

 85a51b7 brings a real performance boost and  006ce65 cuts longer routes down by ~15ms.

### CODE REVIEW

**Comment by TheMarex** [comment]
During this optimization spree I mainly focused on optimizing map matching performance. Key to fast map matching is the coordinate lookup, which used to occupy 85% of the run-time when doing map matching (at least for high-resolution traces).

I used a simple benchmark query on a small dataset, so threat this numbers as relative not real-world values. My test trace went from:

```
49.3284ms/req at 82 coordinate
0.601566ms/coordinate
```

To:
```
26.841ms/req at 82 coordinate
0.327329ms/coordinate
```

The StaticRTree is still the bottle neck here. ~20% are spend in the actual map matching. ~70% in the StaticRTree, the rest is 5% shortest path for the final route and lots of small change. (note this benchmark excluded step generation and geometry generation)

**Comment by oxidase** [comment]
Hi, i have created PR #2348 to make leaf node memory mapped file more system-friendly. util-tests are failing but i can fix it soon.

File: CMakeLists.txt [hunk 1]
```diff
@@ -99,6 +99,7 @@ if(CMAKE_BUILD_TYPE MATCHES Debug)
 endif()
 if(CMAKE_BUILD_TYPE MATCHES Release)
   message(STATUS "Configuring OSRM in release mode")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-omit-frame-pointer")
```
Comment (daniel-j-h) [inline CMakeLists.txt]: Please make sure to remove this before merging.

(I usually set  `CXXFLAGS` in the env for flags like needed by perf, then I don't have to modify the build system)

File: include/engine/routing_algorithms/routing_base.hpp [hunk 2]
```diff
@@ -779,18 +779,50 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
...
+            static_cast<double>(toFloating(target_phantom.location.lon)) * DEGREE_TO_RAD;
+        const double current_cos = std::cos(current_lat);
+
+        const double sin_dlon = std::sin((prev_lon - current_lon) / 2.0);
+        const double sin_dlat = std::sin((prev_lat - current_lat) / 2.0);
+
+        const double aharv = sin_dlat * sin_dlat + prev_cos * current_cos * sin_dlon * sin_dlon;
+        const double charv = 2. * std::atan2(std::sqrt(aharv), std::sqrt(1.0 - aharv));
+        distance += EARTH_RADIUS * charv;
+
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/routing_base.hpp]: You know what's sad? All of this cmath-function heavy code can't be properly optimized as most if not all of those functions potentially set a global `errno` variable. If it turns out that code like this is the bottleneck we might want to move away from legacy cmath functions (in those rare situations where it makes sense!).

(It's also the reason why the cmath functions can't be `constexpr` and thus are not usable at compile time)

File: include/util/coordinate.hpp [hunk 3]
```diff
@@ -181,6 +199,9 @@ bool operator==(const Coordinate lhs, const Coordinate rhs);
 bool operator==(const FloatCoordinate lhs, const FloatCoordinate rhs);
 std::ostream &operator<<(std::ostream &out, const Coordinate coordinate);
 std::ostream &operator<<(std::ostream &out, const FloatCoordinate coordinate);
+
+inline Coordinate::Coordinate(const FloatCoordinate &other) : Coordinate(toFixed(other.lon), toFixed(other.lat)) {}
```
Comment (daniel-j-h) [inline include/util/coordinate.hpp]: If you fwd declare `FloatCoordinate` you can put the function definition into the class where it belongs to and get the inlining for free.

File: include/util/coordinate_calculation.hpp [hunk 4]
```diff
@@ -25,16 +25,49 @@ const constexpr long double EARTH_RADIUS = 6372797.560856;
...
 
 double greatCircleDistance(const Coordinate first_coordinate, const Coordinate second_coordinate);
 
-std::pair<double, FloatCoordinate>
-projectPointOnSegment(const FloatCoordinate &projected_xy_source,
-                      const FloatCoordinate &projected_xy_target,
-                      const FloatCoordinate &projected_xy_coordinate);
+inline std::pair<double, FloatCoordinate> projectPointOnSegment(const FloatCoordinate &source,
+                                                         const FloatCoordinate &target,
+                                                         const FloatCoordinate &coordinate)
```
Comment (daniel-j-h) [inline include/util/coordinate_calculation.hpp]: Should we pass all coordinate types by value? They're wrappers around 2*64bit (`FloatCoordinate`) and 2*32bit (`Coordinate`). Looking at the coordinates why is `FloatLongitude` and `FloatLatitude` `double` anyway, do we really need the precision?

File: include/util/coordinate_calculation.hpp [hunk 5]
```diff
@@ -25,16 +25,49 @@ const constexpr long double EARTH_RADIUS = 6372797.560856;
...
+    const double normed_ratio = unnormed_ratio / squared_length;
+    double clamped_ratio = normed_ratio;
+    if (clamped_ratio > 1.)
+    {
+        clamped_ratio = 1.;
+    }
+    else if (clamped_ratio < 0.)
+    {
+        clamped_ratio = 0.;
+    }
```
Comment (daniel-j-h) [inline include/util/coordinate_calculation.hpp]: I searched the codebase for a `clamp` function and found one in `json_util.hpp` (which is never used by the way :P) --- worth it to provide a utility function for it?

File: include/util/static_rtree.hpp [hunk 6]
```diff
@@ -297,15 +309,17 @@ class StaticRTree
             throw exception("mem index file is empty");
         }
 
-        leaves_stream.open(leaf_file, std::ios::binary);
-        leaves_stream.read((char *)&m_element_count, sizeof(uint64_t));
+        m_leaves_region.open(leaf_file);
```
Comment (daniel-j-h) [inline include/util/static_rtree.hpp]: Handle failure?

File: include/util/static_rtree.hpp [hunk 7]
```diff
@@ -258,6 +265,11 @@ class StaticRTree
         BOOST_ASSERT_MSG(0 < size_of_tree, "tree empty");
         tree_node_file.write((char *)&size_of_tree, sizeof(std::uint32_t));
         tree_node_file.write((char *)&m_search_tree[0], sizeof(TreeNode) * size_of_tree);
+
+        m_leaves_region.open(leaf_node_filename);
```
Comment (daniel-j-h) [inline include/util/static_rtree.hpp]: Handle failure.

File: include/util/static_rtree.hpp [hunk 8]
```diff
@@ -318,8 +332,10 @@ class StaticRTree
             throw exception("mem index file is empty");
         }
 
-        leaves_stream.open(leaf_file, std::ios::binary);
-        leaves_stream.read((char *)&m_element_count, sizeof(uint64_t));
+        m_leaves_region.open(leaf_file);
```
Comment (daniel-j-h) [inline include/util/static_rtree.hpp]: Handle failure? This is three times the same code ^

File: include/util/static_rtree.hpp [hunk 9]
```diff
@@ -491,9 +518,8 @@ class StaticRTree
                 projected_input_coordinate_fixed, projected_nearest);
             // distance must be non-negative
             BOOST_ASSERT(0. <= squared_distance);
-            traversal_queue.push(
-                QueryCandidate{squared_distance, CandidateSegment{Coordinate{projected_nearest},
-                                                                  std::move(current_edge)}});
+            traversal_queue.push(QueryCandidate{squared_distance, SegmentIndex {static_cast<std::uint32_t>(candidate_cache.size())}});
```
Comment (daniel-j-h) [inline include/util/static_rtree.hpp]: Format All The Things

File: src/engine/polyline_compressor.cpp [hunk 10]
```diff
@@ -13,9 +13,8 @@ namespace engine
 namespace /*detail*/ // anonymous to keep TU local
 {
 
-std::string encode(int number_to_encode)
+void encode(int number_to_encode, std::string& output)
```
Comment (daniel-j-h) [inline src/engine/polyline_compressor.cpp]: Why the change here?

Comment (TheMarex) [inline src/engine/polyline_compressor.cpp]: `perf` showed a lot of append operations here. The char-wise append can't be helped but appending the sub-strings just created a lot of unnecessary copies.

File: include/util/coordinate_calculation.hpp (hunk 4 above)
Comment (TheMarex) [inline include/util/coordinate_calculation.hpp]: Using `floats` yields numerical instabilities when doing conversion from/to fixed point representation. There was a bug report on this a while back, especially yielding to problems with high resolution map matched traces (all the coordinates are projected using this function).

We should benchmark if there is actually a difference between passing by value or const ref.

File: include/util/coordinate.hpp (hunk 3 above)
Comment (TheMarex) [inline include/util/coordinate.hpp]: It is forward declared. It just needs to be a complete type when using it here because we access the members. There is a cyclic dependency.

File: include/engine/routing_algorithms/routing_base.hpp (hunk 2 above)
Comment (TheMarex) [inline include/engine/routing_algorithms/routing_base.hpp]: And whta math functions could we use instead? We do spend like 10% of our runtime in `sin` / `cos` / `atan`.

Comment (daniel-j-h) [inline include/engine/routing_algorithms/routing_base.hpp]: There are a couple of constexpr math projects out there, that's why I know about this problem. Never used one of those, probably needs some investigation. I checked Boost.Math and they seem to provide policies for setting errno. But from skimming the source it seems like they're also using cmath in the implementation.

File: include/util/coordinate_calculation.hpp (hunk 5 above)
Comment (TheMarex) [inline include/util/coordinate_calculation.hpp]: They do different things. The json clamp removes values that can't be handled by JSONs number encoding like NaN.

**Comment by TheMarex** [comment]
The functional part of this was broken out in #2385. I would like to move my benchmark to `src/benchmarks` for future reference and then close this.

**Comment by daniel-j-h** [comment]
=> src/CMakeLists.txt :) modularization comes step by step
